ui/app/AppLayouts/Wallet/popups/buy/BuyCryptoModalAdaptor.qml

```diff
@@ -4,143 +4,37 @@ import SortFilterProxyModel 0.2
-    readonly property var selectedToken: selectedTokenEntry.item
-    readonly property var selectedProvider: selectedProviderEntry.item
+    required property var processedTokenSelectorAssetsModel
+    required property var selectedProviderSupportedAssetsArray
+    required property int selectedChainId
 
     readonly property SortFilterProxyModel filteredFlatNetworksModel: SortFilterProxyModel {
         sourceModel: root.networksModel
         filters: ValueFilter { roleName: "isTest"; value: root.areTestNetworksEnabled }
     }
 
-    /* TODO evaluate if this is still needed after
-    https://github.com/status-im/status-desktop/issues/16025 */
-    readonly property ObjectProxyModel plainTokensBySymbolModelWithKey: ObjectProxyModel {
-        sourceModel: root.plainTokensBySymbolModel
-        delegate: SortFilterProxyModel {
-            id: delegateRoot
-            readonly property var addressPerChain: this
-            sourceModel: model.addressPerChain
-            proxyRoles: JoinRole {
-                name: "key"
-                roleNames: ["chainId", "address"]
-                separator: ""
-            }
-            filters: FastExpressionFilter {
-                expression:  !!d.selectedProviderSupportedArray ? d.selectedProviderSupportedArray.includes(model.key) : true
-                expectedRoles: ["key"]
-            }
-        }
-
-        exposedRoles: ["addressPerChain"]
-        expectedRoles: ["addressPerChain"]
-    }
-
-    /* TODO evaluate if this is still needed after
-    https://github.com/status-im/status-desktop/issues/16025 */
-    readonly property ObjectProxyModel groupedAccountAssetsModelWithKey: ObjectProxyModel {
-        sourceModel: root.groupedAccountAssetsModel
-        delegate: SortFilterProxyModel {
-            id: delegateRoot1
-            readonly property var addressPerChain: this
-            sourceModel: model.addressPerChain
-            proxyRoles: JoinRole {
-                name: "key"
-                roleNames: ["chainId", "address"]
-                separator: ""
-            }
-            filters: FastExpressionFilter {
-                expression:  !!d.selectedProviderSupportedArray ? d.selectedProviderSupportedArray.includes(model.key) : true
-                expectedRoles: ["key"]
+    // this proxy removes tokens not supported by selected on-ramp provider
+    readonly property SortFilterProxyModel filteredAssetsModel: SortFilterProxyModel {
+        sourceModel: root.processedTokenSelectorAssetsModel.count ? root.processedTokenSelectorAssetsModel: null
+        filters: FastExpressionFilter {
+            function isSupportedByProvider(addressPerChain) {
+                if(!addressPerChain)
+                    return true
+                return !!ModelUtils.getFirstModelEntryIf(
+                            addressPerChain,
+                            (addPerChain) => {
+                                return root.selectedChainId === addPerChain.chainId &&
```

Comment:
This is heavy! 🙈Even now when switching to Mainnet now there's a small delay. I guess it's fine for now, but we'd probably need a new solution when we'll support lots of tokens.

Reply:
yes true, I think the suggestions from michal below will help simplify this further! Ihave addressed all other comments.